lib/libc/sys/read.2
.\"	$OpenBSD: read.2,v 1.37 2020/02/11 13:19:17 schwarze Exp $
.\"	$NetBSD: read.2,v 1.6 1995/02/27 12:35:47 cgd Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)read.2	8.4 (Berkeley) 2/26/94
.\"
.Dd $Mdocdate: February 11 2020 $
.Dt READ 2
.Os
.Sh NAME
.Nm read ,
.Nm readv ,
.Nm pread ,
.Nm preadv
.Nd read input
.Sh SYNOPSIS
.In unistd.h
.Ft ssize_t
.Fn read "int d" "void *buf" "size_t nbytes"
.Ft ssize_t
.Fn pread "int d" "void *buf" "size_t nbytes" "off_t offset"
.Pp
.In sys/uio.h
.Ft ssize_t
.Fn readv "int d" "const struct iovec *iov" "int iovcnt"
.In sys/types.h
.In sys/uio.h
.Ft ssize_t
.Fn preadv "int d" "const struct iovec *iov" "int iovcnt" "off_t offset"
.Sh DESCRIPTION
.Fn read
attempts to read
.Fa nbytes
of data from the object referenced by the descriptor
.Fa d
into the buffer pointed to by
.Fa buf .
.Fn readv
performs the same action, but scatters the input data into the
.Fa iovcnt
buffers specified by the members of the
.Fa iov
array: iov[0], iov[1], ..., iov[iovcnt-1].
.Fn pread
and
.Fn preadv
perform the same functions, but read from the specified position
.Fa offset
in the file without modifying the file pointer.
.Pp
For
.Fn readv
and
.Fn preadv ,
the
.Fa iovec
structure is defined as:
.Bd -literal -offset indent
struct iovec {
	void *iov_base;
	size_t iov_len;
};
.Ed
.Pp
Each
.Fa iovec
entry specifies the base address and length of an area
in memory where data should be placed.
.Fn readv
and
.Fn preadv
will always fill an area completely before proceeding to the next.
.Pp
On objects capable of seeking, the
.Fn read
starts at a position given by the pointer associated with
.Fa d
(see
.Xr lseek 2 ) .
Upon return from
.Fn read ,
the pointer is incremented by the number of bytes actually read.
.Pp
Objects that are not capable of seeking always read from the current
position.
The value of the pointer associated with such an object is undefined.
.Pp
Upon successful completion,
.Fn read ,
.Fn readv ,
.Fn pread ,
and
.Fn preadv
return the number of bytes actually read and placed in the buffer.
The system guarantees to read the number of bytes requested if
the descriptor references a normal file that has that many bytes left
before the end-of-file, but in no other case.
.Pp
Note that
.Fn readv
and
.Fn preadv
will fail if the value of
.Fa iovcnt
exceeds the constant
.Dv IOV_MAX .
.Sh RETURN VALUES
If successful, the
number of bytes actually read is returned.
Upon reading end-of-file, zero is returned.
Otherwise, a \-1 is returned and the global variable
.Va errno
is set to indicate the error.
.Sh ERRORS
.Fn read ,
.Fn readv ,
.Fn pread ,
and
.Fn preadv
will fail if:
.Bl -tag -width Er
.It Bq Er EBADF
.Fa d
is not a valid file or socket descriptor open for reading.
.It Bq Er EFAULT
Part of
.Fa buf
points outside the process's allocated address space.
.It Bq Er EINTR
A read from a slow device
(i.e. one that might block for an arbitrary amount of time)
was interrupted by the delivery of a signal
before any data arrived.
.It Bq Er EIO
An I/O error occurred while reading from the file system.
.It Bq Er EISDIR
The underlying file is a directory.
.El
.Pp
In addition,
.Fn read
and
.Fn readv
may return the following errors:
.Bl -tag -width Er
.It Bq Er EAGAIN
The file was marked for non-blocking I/O,
and no data were ready to be read.
.It Bq Er ENOTCONN
The file is a socket associated with a connection-oriented protocol
and has not been connected.
.It Bq Er EIO
The process is a member of a background process attempting to read
from its controlling terminal, the process is ignoring or blocking
the
.Dv SIGTTIN
signal or the process group is orphaned.
.El
.Pp
.Fn read
and
.Fn pread
may return the following error:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa nbytes
was larger than
.Dv SSIZE_MAX .
.El
.Pp
.Fn pread
and
.Fn preadv
may return the following errors:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa offset
was negative.
.It Bq Er ESPIPE
.Fa d
is associated with a pipe, socket, FIFO, or tty.
.El
.Pp
.Fn readv
and
.Fn preadv
may return the following errors:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa iovcnt
was less than or equal to 0, or greater than
.Dv IOV_MAX .
.It Bq Er EINVAL
The sum of the
.Fa iov_len
values in the
.Fa iov
array overflowed an
.Vt ssize_t .
.It Bq Er EFAULT
Part of
.Fa iov
points outside the process's allocated address space.
.El
.Sh SEE ALSO
.Xr dup 2 ,
.Xr fcntl 2 ,
.Xr open 2 ,
.Xr pipe 2 ,
.Xr poll 2 ,
.Xr select 2 ,
.Xr socket 2 ,
.Xr socketpair 2
.Sh STANDARDS
The
.Fn read ,
.Fn readv ,
and
.Fn pread
functions conform to
.St -p1003.1-2008 .
.Sh HISTORY
A
.Fn read
system call first appeared in
.At v1 ;
.Fn readv
in
.Bx 4.1c ;
.Fn pread
in
.At V.4 ;
and
.Fn preadv
in
.Ox 2.7 .
.Sh CAVEATS
Error checks should explicitly test for \-1.
Code such as
.Bd -literal -offset indent
while ((nr = read(fd, buf, sizeof(buf))) > 0)
.Ed
.Pp
is not maximally portable, as some platforms allow for
.Fa nbytes
to range between
.Dv SSIZE_MAX
and
.Dv SIZE_MAX
\- 2, in which case the return value of an error-free
.Fn read
may appear as a negative number distinct from \-1.
Proper loops should use
.Bd -literal -offset indent
while ((nr = read(fd, buf, sizeof(buf))) != -1 && nr != 0)
.Ed


usr.bin/tail/read.c
/*	$OpenBSD: read.c,v 1.20 2017/03/26 19:55:07 martijn Exp $	*/
/*	$NetBSD: read.c,v 1.4 1994/11/23 07:42:07 jtc Exp $	*/

/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Edward Sze-Tyan Wang.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "extern.h"

/*
 * bytes -- read bytes to an offset from the end and display.
 *
 * This is the function that reads to a byte offset from the end of the input,
 * storing the data in a wrap-around buffer which is then displayed.  If the
 * rflag is set, the data is displayed in lines in reverse order, and this
 * routine has the usual nastiness of trying to find the newlines.  Otherwise,
 * it is displayed from the character closest to the beginning of the input to
 * the end.
 *
 * A non-zero return means an (non-fatal) error occurred.
 *
 */
int
bytes(struct tailfile *tf, off_t off)
{
	int ch;
	size_t len, tlen;
	char *ep, *p, *t;
	int wrap;
	char *sp;

	if (off > SIZE_MAX)
		errx(1, "offset too large");

	if ((sp = p = malloc(off)) == NULL)
		err(1, NULL);

	for (wrap = 0, ep = p + off; (ch = getc(tf->fp)) != EOF;) {
		*p = ch;
		if (++p == ep) {
			wrap = 1;
			p = sp;
		}
	}
	if (ferror(tf->fp)) {
		ierr(tf->fname);
		free(sp);
		return(1);
	}

	if (rflag) {
		for (t = p - 1, len = 0; t >= sp; --t, ++len)
			if (*t == '\n' && len) {
				WR(t + 1, len);
				len = 0;
			}
		if (wrap) {
			tlen = len;
			for (t = ep - 1, len = 0; t >= p; --t, ++len)
				if (*t == '\n') {
					if (len) {
						WR(t + 1, len);
						len = 0;
					}
					if (tlen) {
						WR(sp, tlen);
						tlen = 0;
					}
				}
			if (len)
				WR(t + 1, len);
			if (tlen)
				WR(sp, tlen);
		}
	} else {
		if (wrap && (len = ep - p))
			WR(p, len);
		if ((len = p - sp))
			WR(sp, len);
	}

	free(sp);
	return(0);
}

/*
 * lines -- read lines to an offset from the end and display.
 *
 * This is the function that reads to a line offset from the end of the input,
 * storing the data in an array of buffers which is then displayed.  If the
 * rflag is set, the data is displayed in lines in reverse order, and this
 * routine has the usual nastiness of trying to find the newlines.  Otherwise,
 * it is displayed from the line closest to the beginning of the input to
 * the end.
 *
 * A non-zero return means an (non-fatal) error occurred.
 *
 */
int
lines(struct tailfile *tf, off_t off)
{
	struct {
		size_t blen;
		size_t len;
		char *l;
	} *lines = NULL;
	int ch, rc = 0;
	char *p = NULL;
	int wrap;
	size_t cnt, lineno, nlineno, recno, blen, newsize;
	char *sp = NULL, *newp = NULL;

	if (off > SIZE_MAX)
		errx(1, "offset too large");

	lineno = blen = cnt = recno = wrap = 0;

	while ((ch = getc(tf->fp)) != EOF) {
		if (++cnt > blen) {
			newsize = blen + 1024;
			if ((newp = realloc(sp, newsize)) == NULL)
				err(1, NULL);
			sp = newp;
			blen = newsize;
			p = sp + cnt - 1;
		}
		*p++ = ch;
		if (recno >= lineno) {
			nlineno = (lineno + 1024) > off ?
			    (size_t) off : lineno + 1024;
			if ((lines = recallocarray(lines, lineno, nlineno,
			    sizeof(*lines))) == NULL)
				err(1, NULL);
			lineno = nlineno;
		}
		if (ch == '\n') {
			if (lines[recno].blen < cnt) {
				newsize = cnt + 256;
				if ((newp = realloc(lines[recno].l,
				    newsize)) == NULL)
					err(1, NULL);
				lines[recno].l = newp;
				lines[recno].blen = newsize;
			}
			memcpy(lines[recno].l, sp, (lines[recno].len = cnt));
			cnt = 0;
			p = sp;
			if (++recno == off) {
				wrap = 1;
				recno = 0;
			}
		}
	}
	if (ferror(tf->fp)) {
		ierr(tf->fname);
		rc = 1;
		goto done;
	}
	if (cnt) {
		lines[recno].l = sp;
		lines[recno].len = cnt;
		sp = NULL;
		if (++recno == off) {
			wrap = 1;
			recno = 0;
		}
	}

	if (rflag) {
		for (cnt = recno; cnt > 0; --cnt)
			WR(lines[cnt - 1].l, lines[cnt - 1].len);
		if (wrap)
			for (cnt = off; cnt > recno; --cnt)
				WR(lines[cnt - 1].l, lines[cnt - 1].len);
	} else {
		if (wrap)
			for (cnt = recno; cnt < off; ++cnt)
				WR(lines[cnt].l, lines[cnt].len);
		for (cnt = 0; cnt < recno; ++cnt)
			WR(lines[cnt].l, lines[cnt].len);
	}
done:
	for (cnt = 0; cnt < lineno; cnt++)
		free(lines[cnt].l);
	free(sp);
	free(lines);
	return(rc);
}
